File: read_bbtfile.py
import pandas as pd
import numpy as np
from datetime import datetime

def read_bbtfile(filepath):
    # Read the file in CSV format
    data = pd.read_csv(filepath, header=None).to_numpy()

    # Extract filename
    filename = filepath.split('/')[-1].split('.')[0]

    SA = {'Shot': []}
    si = 0

    shot = {
        'Route': [],
        'Route0': [],
        'hit': 0
    }

    for bi in range(3):
        route = {
            't': data[:, 0] / 1000,
            'x': data[:, (bi * 2) + 1],
            'y': data[:, (bi * 2) + 2]
        }

        # Calculate movement distance
        dL = np.sqrt((route['x'] - route['x'][0])**2 + (route['y'] - route['y'][0])**2)
        ind = np.where(dL > 5)[0]

        if ind.size == 0:
            route['t'] = np.delete(route['t'], slice(1, -1))
            route['x'] = np.delete(route['x'], slice(1, -1))
            route['y'] = np.delete(route['y'], slice(1, -1))
        else:
            route['t'] = np.delete(route['t'], slice(1, ind[0] + 1))
            route['x'] = np.delete(route['x'], slice(1, ind[0] + 1))
            route['y'] = np.delete(route['y'], slice(1, ind[0] + 1))

        shot['Route'].append(route)
        shot['Route0'].append(route)

    SA['Shot'].append(shot)

    SA['Table'] = pd.DataFrame({
        'Selected': [False],
        'ShotID': [int(datetime.now().strftime('%y%m%d%H%M%S%f'))],
        'Mirrored': [0],
        'Filename': [filename],
        'GameType': ['3'],
        'Interpreted': [0],
        'Player': [0],
        'ErrorID': [-1],
        'ErrorText': ['only read in']
    })

    return SA

File: read_BBTfile.m
function SA = read_BBTfile(filepath)

% Read the file csv format
data = readmatrix(filepath, 'FileType', 'text');

% extract filename
[~,filename,~] = fileparts(filepath);

si = 0;
SA.Shot = [];
check = 0;

si = si + 1;
Selected{si,1} = false;
Filename{si,1} = filename;
GameType{si,1} = '3';
ShotID{si,1} = str2num(char(datetime('now','Format','yyMMddHHmmssSSS')));
Player{si,1} = 0;

ErrorID{si,1} = -1;
ErrorText{si,1} = 'only read in';
Interpreted{si,1} = 0;
Mirrored{si,1} = 0;

for bi = 1:3
    SA.Shot(si).Route(bi).t = data(:,1)/1000;
    SA.Shot(si).Route(bi).x = data(:,(bi-1)*2 + 2);
    SA.Shot(si).Route(bi).y = data(:,(bi-1)*2 + 3);
    % Calculate movement distance
    dL = sqrt((SA.Shot(si).Route(bi).x-SA.Shot(si).Route(bi).x(1)).^2 + (SA.Shot(si).Route(bi).y-SA.Shot(si).Route(bi).y(1)).^2);

    ind = find(dL > 5);
    if isempty(ind)
        % Ball didn't move. Only t1 end Tlast is given
        SA.Shot(si).Route(bi).t(2:end-1) = [];
        SA.Shot(si).Route(bi).x(2:end-1) = [];
        SA.Shot(si).Route(bi).y(2:end-1) = [];
    else
        SA.Shot(si).Route(bi).t([2:ind(1)]) = [];
        SA.Shot(si).Route(bi).x([2:ind(1)]) = [];
        SA.Shot(si).Route(bi).y([2:ind(1)]) = [];
    end
    
%    SA.Shot(si).Route(bi) = correct_velocity(SA.Shot(si).Route(bi));

    SA.Shot(si).Route0(bi) = SA.Shot(si).Route(bi);
    check = 1;
end
SA.Shot(si).hit=0;

if check
    SA.Table = table(Selected, ShotID, Mirrored, Filename, GameType, Interpreted, Player, ErrorID, ErrorText);
else
    SA = [];
end

